### DISCUSSION TITLE
Teleport SSH in high latency scenarios & IOT

### DISCUSSION THREAD

**Discussion by craigbryson** [gravitational/teleport#27801] [Q&A]
Hi!

I am testing teleport to connect via SSH to IoT nodes that have connections with high latency at times. When joining a node like this I see the following in the agent logs

```
Jun 13 14:17:24 my-node teleport[167688]: 2023-06-13T14:17:24Z INFO [PROC:1]    Joining the cluster with a secure token. pid:167688.1 service/connect.go:589
Jun 13 14:17:24 my-node teleport[167688]: 2023-06-13T14:17:24Z INFO [AUTH]      Attempting registration via proxy server. auth/register.go:277
Jun 13 14:17:24 my-node teleport[167688]: 2023-06-13T14:17:24Z INFO [AUTH]      Attempting registration via proxy server. auth/register.go:277
Jun 13 14:17:31 my-node teleport[167688]: 2023-06-13T14:17:31Z INFO [AUTH]      Successfully registered via proxy server. auth/register.go:284
Jun 13 14:17:31 my-node teleport[167688]: 2023-06-13T14:17:31Z INFO [PROC:1]    Node has obtained credentials to connect to the cluster. pid:167688.1 service/connect.go:646
Jun 13 14:17:31 my-node teleport[167688]: 2023-06-13T14:17:31Z INFO [AUTH]      Successfully registered via proxy server. auth/register.go:284
Jun 13 14:17:31 my-node teleport[167688]: 2023-06-13T14:17:31Z INFO [PROC:1]    Instance has obtained credentials to connect to the cluster. pid:167688.1 service/connect.go:646
Jun 13 14:17:59 my-node teleport[167688]: 2023-06-13T14:17:59Z ERRO [PROC:1]    Node failed to establish connection to cluster: Failed to connect to Proxy Server through tunnel: connection error: desc = "transport: authentication handshake failed: context deadline exceeded". pid:167688.1 service/connect.go:123
Jun 13 14:17:59 my-node teleport[167688]: 2023-06-13T14:17:59Z ERRO [PROC:1]    Instance failed to establish connection to cluster: Failed to connect to Proxy Server through tunnel: connection error: desc = "transport: authentication handshake failed: context deadline exceeded". pid:167688.1 service/connect.go:123
```

I know that unreliable connectivity is not ideal but I wonder if anyone has experience using Teleport in such cases? Or if there's anything I can tweak to make the join more likely to succeed?

Thanks,

Craig

**Comment by webvictim** [comment]
`context deadline exceeded` is essentially a timeout. Unfortunately Teleport does have a lot of back-and-forth traffic between nodes and the Teleport proxy/auth servers to establish its tunnels, meaning that latency is going to be a key factor in whether Teleport is able to successfully establish a tunnel or not. This is especially true during a first-time node registration.

A lot of Teleport's internal operations are based on a 30 second timeout which seems roughly in line with the issue you're seeing. Those timeouts are set here:

https://github.com/gravitational/teleport/blob/583d0d32489239ae5ccc08fb6ee56f9f6aa2b63b/api/defaults/defaults.go#L31-L35

As such, you could try cloning the Teleport Github repo, modifying those timeouts to a higher value (like 60, 90 or 120 seconds) then recompiling the Teleport binary and using that on your remote nodes instead. The extra time might be enough for the initial join operation to complete. Once that's done, ongoing operations will use a TLS client certificate which is cached on disk on the node, so should be a more "condensed" operation at high latencies.

You can recompile the binary by running `make build/teleport` in the Github checkout, assuming that you don't need support for PAM, BPF or other optional features. You'll need Go 1.20+ installed plus `make` and a number of other dependencies. You can `make -C build.assets build-binaries` to compile in Docker if you prefer (which will already have all the necessary dependencies), but it will take longer.

**Reply by craigbryson** [reply]
Thanks @webvictim, I was away for a few days but I'll try this today and report back

**Reply by craigbryson** [reply]
Hey,

I tried to build using docker from both the master branch as a base and 13.1 branch but I encountered the error below, is there a specific branch I should base my patch branch on?

```
... (omitted a lot of successful lines)
 github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/sql/armsql: /opt/go/pkg/tool/linux_amd64/compile: signal: killed
github.com/aws/aws-sdk-go-v2/service/ec2: /opt/go/pkg/tool/linux_amd64/compile: signal: killed
# github.com/aquasecurity/libbpfgo
In file included from /tmp/gomodcache/github.com/aquasecurity/libbpfgo@v0.4.5-libbpf-1.0.1/libbpfgo.h:16:0,
                 from /tmp/gomodcache/github.com/aquasecurity/libbpfgo@v0.4.5-libbpf-1.0.1/libbpfgo.go:5:
/usr/libbpf-1.0.1/include/bpf/libbpf.h:70:54: warning: ‘enum bpf_link_type’ declared inside parameter list [enabled by default]
 LIBBPF_API const char *libbpf_bpf_link_type_str(enum bpf_link_type t);
                                                      ^
/usr/libbpf-1.0.1/include/bpf/libbpf.h:70:54: warning: its scope is only this definition or declaration, which is probably not what you want [enabled by default]
make[2]: *** [build/teleport] Error 1
make[2]: Leaving directory `/go/src/github.com/gravitational/teleport'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/go/src/github.com/gravitational/teleport'
make: *** [full] Error 2
make: Leaving directory `/go/src/github.com/gravitational/teleport'
make: *** [build-binaries] Error 2
```

**Reply by webvictim** [reply]
Hmm, building from the `v13.1.1` tag and applying your changes on top of that should work fine.

**Reply by craigbryson** [reply]
Hi! I forked again just to be sure I was using the right tag but I get the same error building in docker :(. Do you have any other suggestions? I will keep trying

**Reply by craigbryson** [reply]
Hi @webvictim - scratch that, one of my colleagues managed to build it, however, still seeing the same.

We had a look through the code and I think we're hitting the `process.Config.ClientTimeout` in `func (process *TeleportProcess) newClientThroughTunnel()`. As far as I can see this isn't influenced by the defaults you mentioned? Do you have an idea where it should be set?

**Reply by rosstimothy** [reply]
@webvictim is correct to extend the timeout you should only have to change `DefaultIOTimeout`. Both the SSH connection to the Proxy and the connection to Auth through the tunnel use that value.
